blog/_src/posts/2017-11-XX-adding-refinement-types.md

```diff
@@ -0,0 +1,449 @@
+                A)))
+(define safe-vector-ref unsafe-vector-ref)
+
+(: safe-vector-set!
+   (All (A) (-> ([v : (Vectorof A)]
+                 [n : Natural]
+                 [a : A])
+                #:pre (v n) (< n (vector-length v))
+                Void)))
+(define safe-vector-set! unsafe-vector-set!)
+```
+
+Now that we have these foundational operations in place
+(i.e. `make-vector`, `safe-vector-ref`, and `safe-vector-set!`) ,
+let's see if we can define some of the [vector
+operations](https://docs.racket-lang.org/reference/vectors.html) that
+are provided by `racket/base` and `racket/vector`, but with more
+precise types!
+
+### `build-vector`
+
+Let's look at the documentation for `build-vector`:
+
+```
+(build-vector n proc) → vector?
+  n : exact-nonnegative-integer?
+  proc : (exact-nonnegative-integer? . -> . any/c)
+
+Creates a vector of n elements by applying proc to the
+integers from 0 to (sub1 n) in order. If vec is the 
+resulting vector, then (vector-ref vec i) is the value 
+produced by (proc i).
+
+Example:
+> (build-vector 5 add1)
+'#(1 2 3 4 5)
+```
+
+First let's consider its type signature. It seems like there are two places where we could use refinement types to create a more precise type-based specification:
+
+1. the `proc` procedure will always be given an integer in the range
+[0, `n`), and
+
+2. the length of the returned vector is equal to `n`.
+
+We can encode both of these directly in the type of `build-vector`:
+
+```racket
+(: build-vector
+   (All (A) (-> ([n : Natural]
+                 [proc : (n) (-> (Refine [i : Natural] (< i n)) A)])
+                (Refine [v : (Vectorof A)]
+                        (= n (vector-length v))))))
+```
+
+With the type signature in place, we can proceed to define the
+function:
+
+```racket
+(define (build-vector n proc)
+  (cond
+    [(> n 0)
+     (define init-val (proc 0))
+     (define vec (make-vector n init-val))
+     (let loop! ([i : Natural 1])
+       (when (< i n)
+         (safe-vector-set! vec i (proc i))
+         (loop! (add1 i))))
+     vec]
+    [else (vector)]))
+```
+
+There are a few things to note about the definition:
+
+a. In order to make a `(Vectorof A)`, we need some value of type `A`
+to initialize the vector with (i.e. the second arg to
+`make-vector`). Because of this, we first check if `n` is greater than
+0. If it is greater than 0, we can compute `(proc 0)` and use that to
+initialize the vector we create, otherwise we simply return `(vector)`
+(i.e. an empty vector) which is "vacuously" a vector of `A` for any
+`A`.
+
+b. The recursive loop we use to populate indices 1 through `(sub1 n)`,
+which we named `loop!`, requires a type annotation `Natural` on its
+argument `i` so that its being greater than or equal to 0 is known in
+the body of the recursive function. That, combined with the test `(< i
+n)`, allows us to successfully use `safe-vector-set!` to update the
+contents at `i`. (Note: if Typed Racket's type inference were not
```

Comment:
Note: if Typed Racket's type inference were smarter here, this annotation would not be necessary.